deeppavlov/core/models/tf_model.py

```diff
@@ -13,119 +13,123 @@
 See the License for the specific language governing permissions and
 limitations under the License.
 """
-
 import sys
 from abc import abstractmethod
+from collections import defaultdict
+import numpy as np
 
 import tensorflow as tf
-from overrides import overrides
 
 from deeppavlov.core.models.nn_model import NNModel
-from deeppavlov.core.models.serializable import Serializable
-from deeppavlov.core.common.attributes import check_attr_true
 from deeppavlov.core.common.log import get_logger
-from deeppavlov.core.common.errors import ConfigError
 from .tf_backend import TfModelMeta
 
-"""
-Here is an abstract class for neural network models based on Tensorflow.
-If you use something different, ex. Pytorch, then write similar to this class, inherit it from
-Trainable and Inferable interfaces and make a pull-request to deeppavlov.
-"""
-
 log = get_logger(__name__)
 
 
-class TFModel(Serializable, metaclass=TfModelMeta):
-    def __init__(self, **kwargs):
-        self._saver = tf.train.Saver
-        super().__init__(**kwargs)
+class TFModel(NNModel, metaclass=TfModelMeta):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
     @abstractmethod
-    def _add_placeholders(self):
-        """
-        Add all needed placeholders for a computational graph.
-        """
-        pass
+    def train_on_batch(self, x_batch, y_batch):
+        """ Perform single step of optimization given batch of samples
 
-    @abstractmethod
-    def run_sess(self, *args, **kwargs):
-        """
-        1. Call _build_graph()
-        2. Define all comuptations.
-        3. Run tf.sess.
-        3. Reset state if needed.
-        :return:
-        """
-        pass
+        Args:
+            x_batch: batch of x-s it could be anything: dict, list, numpy array. However,
+                     it must contain  number of samples
+            y_batch: batch of y-s. It must contain same number of samples as x_batch.
+
+        Returns:
+            loss: mean loss over batch
 
-    @abstractmethod
-    def _train_step(self, features, *args):
-        """
-        Define a single training step. Feed dict to tf session.
-        :param features: input features
-        :param args: any other inputs, including target vector, you need to pass for training
-        :return: metric to return, usually loss
         """
         pass
 
+    def load(self):
+        """Load model parameters from self.load_path"""
+        path = str(self.load_path.resolve())
```

Comment:
Why would you need to resolve save and load paths here? Are they not resolved in the Serializable class?

Reply:
save/load pathes are user_expanded but not resolved:
```def expand_path(path: Union[str, Path]) -> Path:
    return get_deeppavlov_root() / Path(path).expanduser()```